types/src/keyless/proof_simulation.rs

```diff
@@ -0,0 +1,334 @@
+use ark_groth16::{prepare_verifying_key, Groth16};
+use ark_crypto_primitives::snark::SNARK;
+use ark_ec::pairing::Pairing;
+use ark_ff::Field;
+use ark_relations::r1cs::{ConstraintSynthesizer, SynthesisError};
+use ark_std::{
+    rand::{RngCore, SeedableRng},
+    test_rng, UniformRand,
+};
+use ark_std::{marker::PhantomData, vec::Vec};
+use ark_groth16::r1cs_to_qap::{LibsnarkReduction, R1CSToQAP};
+use ark_groth16::data_structures::{VerifyingKey, Proof};
+use ark_serialize::*;
+use ark_std::rand::Rng;
+use ark_relations::r1cs::Result as R1CSResult;
+use ark_ec::{AffineRepr,CurveGroup};
+use ark_ff::PrimeField;
+use std::ops::AddAssign;
+use ark_bn254::Bn254;
+use ark_circom::CircomConfig;
+use ark_circom::CircomBuilder;
+use std::fs::File;
+use std::collections::HashMap;
+use std::str::FromStr;
+use num_bigint_v4::BigInt;
+use ark_ff::MontBackend;
+use ark_bn254::FrConfig;
+use crate::keyless::bn254_circom::{g1_projective_str_to_affine, g2_projective_str_to_affine};
+
+/// The SNARK of [[Groth16]](https://eprint.iacr.org/2016/260.pdf), where "proving" implements the
+/// simulation algorithm instead, using the trapdoor output by the modified setup algorithm also
+/// implemented in this struct
+pub struct Groth16Simulator<E: Pairing, QAP: R1CSToQAP = LibsnarkReduction> {
+    _p: PhantomData<(E, QAP)>,
+}
+
+/// The simulation prover key for for the Groth16 zkSNARK, used only for simulating proofs with the
+/// secret trapdoor information
+#[derive(Clone, Debug, PartialEq, CanonicalSerialize, CanonicalDeserialize)]
+pub struct ProvingKeyWithTrapdoor<E: Pairing> {
+    /// Vector of elements from the verifying key
+    pub gamma_abc_g1: Vec<E::G1Affine>,
+    /// Trapdoor alpha
+    pub alpha: E::ScalarField,
+    /// Trapdoor beta
+    pub beta: E::ScalarField,
+    /// Trapdoor delta
+    pub delta: E::ScalarField,
+    /// Trapdoor gamma 
+    pub gamma: E::ScalarField,
+    /// Generator for G1
+    pub g1: E::G1Affine,
+    /// Generator for G2
+    pub g2: E::G2Affine,
+}
+
+impl<E: Pairing, QAP: R1CSToQAP> Groth16Simulator<E, QAP> 
+{
+    fn circuit_agnostic_specific_setup_with_trapdoor<R: RngCore>(
```

Comment:
`circuit_agnostic_specific_setup_with_trapdoor` should be `circuit_agnostic_setup_with_trapdoor`